contour-lint/src/main/java/com/squareup/contour/lint/NestedContourLayoutsDetector.kt

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright 2020 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.squareup.contour.lint
+
+import com.android.tools.lint.client.api.UElementHandler
+import com.android.tools.lint.detector.api.Category
+import com.android.tools.lint.detector.api.Detector
+import com.android.tools.lint.detector.api.Implementation
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.JavaContext
+import com.android.tools.lint.detector.api.LintFix
+import com.android.tools.lint.detector.api.Scope
+import com.android.tools.lint.detector.api.Severity.ERROR
+import com.intellij.psi.PsiClass
+import com.intellij.psi.PsiClassType
+import com.intellij.psi.PsiType
+import org.jetbrains.kotlin.psi.psiUtil.getStartOffsetIn
+import org.jetbrains.uast.UCallExpression
+import org.jetbrains.uast.UElement
+import org.jetbrains.uast.ULambdaExpression
+import org.jetbrains.uast.getContainingUClass
+import org.jetbrains.uast.getParentOfType
+import org.jetbrains.uast.getUCallExpression
+
+@Suppress("UnstableApiUsage")
+class NestedContourLayoutsDetector : Detector(), Detector.UastScanner {
+
+  companion object {
+    val ISSUE = Issue.create(
+        id = "NestedContourLayouts",
+        briefDescription = "Incorrectly nested ContourLayouts",
+        explanation = """
+          When a nested contour view is initialized and laid out using a lambda that offers \
+          `this` as an argument (e.g., Kotlin's `apply{}` and `run{}`), it's easy to accidentally \
+          call `layout()`/`applyLayout()` on the wrong scope. This lint flags these kinds of \
+          errors. Here's an example:
+          
+          ```
+          class FooView : ContourLayout {
+            val view = BarView(context).apply {
+              layout(x, y) <- ERROR! The receiver of this call is BarView and not FooView!
+            }
+          }
+          class BarView(context: Context) : ContourLayout
+          ```
+          
+          The above example result in a StackOverflowException because `FooView` ends up laying \ 
+          out itself rather than `BarView`.
+        """.trimMargin(),
+        category = Category.CORRECTNESS,
+        severity = ERROR,
+        priority = 5,
+        implementation = Implementation(
+            NestedContourLayoutsDetector::class.java,
+            Scope.JAVA_FILE_SCOPE
+        )
+    )
+  }
```

Comment:
nit: move companion object to bottom of source file

Reply:
done